### ISSUE TITLE
Wrong example in authenticateToken for SimplePreAuthenticationListener

### ISSUE THREAD

**Issue by s7anley** [symfony/symfony-docs#4148]
In 2nd code example in [storing authentication in the session section](http://symfony.com/doc/current/cookbook/security/api_key_authentication.html#storing-authentication-in-the-session), it's not possible to get current User with `$user = $token->getUser();`. SimplePreAuthenticationListener is skipped when current user is not authenticate with `AnonymousToken`, which in this case is true.

Or maybe I'm missing something?
Thanks.

**Comment by PapyDanone** [comment]
hey @peterrehm in your attempt to reproduce this issue does it still work if you remove `anonymous: ~` from your firewall?

I'm actually getting an exception on 2nd request too and can somewhat relate to what @s7anley was saying. 

Here's my security.yml:

```
secured_area:
     pattern: ^/
     stateless: false
     simple-preauth:
           authenticator: api_key_authenticator
```

The exception:

    Uncaught PHP Exception Symfony\Component\Security\Core\Exception\AuthenticationCredentialsNotFoundException: "A Token was not found in the SecurityContext." at /Users/jeromeguilbot/Sites/tma-api/vendor/symfony/symfony/src/Symfony/Component/Security/Http/Firewall/AccessListener.php line 53


In my code, the only thing that differs from the example is, in `ApiKeyAuthenticator.php`:

```
public function createToken(Request $request, $providerKey)
{
        // look for an apikey query parameter
        $apiKey = $request->query->get('apikey');

        // or if you want to use an "apikey" header, then do something like this:
        // $apiKey = $request->headers->get('apikey');

        if (!$apiKey) {
            return null;
        }

        return new PreAuthenticatedToken(
            'anon.',
            $apiKey,
            $providerKey
        );
}
```

Basically I want the user to be either authenticated by an API key in the url **or** a token in the session if it is found, and that for all requests.

Am I missing something?

**Comment by peterrehm** [comment]
You can omit the anonymous setting, I needed it since you can login either by a form or by the api key. The application supports both.

But if I recall correctly you should not return null, just return an empty apiKey in the PreAuthenticated Token or 

````php
public function createToken(Request $request, $providerKey)
{
        // look for an apikey query parameter
        $apiKey = $request->query->get('apikey');

        // or if you want to use an "apikey" header, then do something like this:
        // $apiKey = $request->headers->get('apikey');

        if (!$apiKey) {
            throw new BadCredentialsException('No Auth token found');
        }

        return new PreAuthenticatedToken(
            'anon.',
            $apiKey,
            $providerKey
        );
}
````

I assume based on the provided error message you somehow have no apiKey and therefore no token is returned.

**Comment by PapyDanone** [comment]
Thanks for your quick response.

Indeed, I only pass the API key as a parameter in first request.

If, as you suggest, I return an empty apiKey in the PreAuthenticated token as follows:
```php
if (!$apiKey) {
    return new PreAuthenticatedToken(
           'anon.',
           null,
           $providerKey
    );
}
```
I get the `Authentication Failed. API Key "" does not exist.` error thrown in the `authenticateToken` method. 

As @s7anley was saying, `$user = $token->getUser();` doesn't work on the 2nd pass (it contains the string `anon.` in my case, since it is what gets returned in `createToken()`). Also note that `createToken` is indeed called for every subsequent request, contrary to what you seemed to say earlier: "On the next request however the createToken() will not be called any more".

Does that clarify the situation a bit?

**Comment by peterrehm** [comment]
Actually not really. IIRC createToken must not be called it there is the user/token in the session.
It looks like your token is not being stored to the session. It must be related to that. Have you found
out some more informations?

**Comment by PapyDanone** [comment]
I think I went to the bottom of it. 

In my case the `refreshUser()` method from `ApiKeyUserProvider.php` was never called on the second request. Instead, it was trying to call the `refreshUser()` of the default `in_memory` provider that you get with a fresh Symfony install. The user not being found, it would then completely invalidate the token with `$this->context->setToken($token)`, `$token` being null; 

Replacing the default provider with:
```yml
security:
    providers:
        simple_preauth_provider:
            id: api_user_provider
``` 
with `api_user_provider` referencing the service defined in the example, the whole thing finally works. I'm guessing it's been working for you @peterrehm all along because you had `provider: fos_userbundle` defined as the user provider, which would have reloaded the user from the database every time. 

So I think we should simply update the docs with the snippet above... Am I missing something?

**Comment by peterrehm** [comment]
I just had a look at it and must admit you are right. Registering the user provider has been abstracted in the article with just a short reference to the corresponding doc section about registering a custom user provider. 

If you have no registered user provider it breaks obviously.

**Comment by peterrehm** [comment]
Just added a PR.

**Comment by PapyDanone** [comment]
Thanks for updating this @peterrehm

**Comment by peterrehm** [comment]
Thanks for investigating it :)
